```{r}
library(tidyverse)
library(qs)
library(Seurat)
library(reshape2)
library(future)
library(limma)
library(edgeR)
library(sva)
library(fgsea)
library(furrr)
library(flextable)
library(ggrepel)
library(cowplot)
library(egg)
library(ComplexHeatmap)
library(circlize)
library(scales)
```

## Set up directories
```{r}
base_dir = "~/nest/rstudio" ## to change as needed
prefix = file.path(base_dir, "neuropeptide/SLCR")
prefix_data = file.path(prefix, "deaf_bus_trimmed_merge_3p_seurat_lman-filter_prep")
script_name = "02-deaf_bus_trimmed_merge_3p_seurat_lman-filter_kl-mean_frac-mito_block-tags_with-sva_gene-filt"
prefix_cur = file.path(prefix_data, script_name)
dir.create(prefix_cur, recursive = T)
```

```{r}
util_dir = file.path(base_dir, "utils")
source(file.path(util_dir, "common_aesthetics.R"))
```

## Load neuropeptide genes
```{r}
nps = read_delim("~/nest/gene_lists/neuropeptide_related_genes.txt")
```


## Function def
```{r}
run_fit = function(se, design_char, blocking) {
  
  info_df1 = se@meta.data
  info_df1 = droplevels(info_df1)
  
  dat_cur = GetAssayData(se, "counts", assay="RNA")
  dge = DGEList(dat_cur)
  
  ### Filter genes
  to_keep = filterByExpr(dge, group = info_df1[,blocking])
  sprintf("Num genes to keep: %s", sum(to_keep))
  dge = dge[to_keep,]
  
  dge = calcNormFactors(dge,method =c("TMMwsp"))
  
  ### FIT 
  design = model.matrix(as.formula(design_char), info_df1)
  colnames(design) = make.names(colnames(design))
  print(colnames(design))
  v1 = voom(dge,design,plot=F, normalize.method=normalize.method)
  corfit = duplicateCorrelation(v1, design, block = info_df1[,blocking])
  fit = lmFit(v1, design, block = info_df1[,blocking], correlation =
                corfit$consensus.correlation)
  return(fit)
}
```

## Load object
```{r}
se_fname = file.path(prefix_data, "seurat_object.qs")
se1 = qread(se_fname)
```

## Load coefs
```{r}
model_fname = file.path(prefix_data, "coefs.rds")
se_coefs = readRDS(model_fname)
```

```{r}
loc_trans = read_delim("~/nest/assembly/lonStrDom2/ncbi/GCF_005870125.1_lonStrDom2_genomic_loc_synonyms.txt", delim="\t")
```

## Gene filter
Remove genes that have similar expression variation across individuals and regions -- likely influenced by technical artifact
```{r}

redo = F
gene_var_fname = file.path(prefix_data, "gene_variance_stats.rds")
if (redo) {
  dat = GetAssayData(se1, assay = "RNA", slot="data")
  
  md = FetchData(se1, c("set", "duration.of.experiment", "procedure2", "tags", "position", "index2", "nCount_RNA")) %>%
    rownames_to_column("id")
  
  dat_df_full = dat %>% as.data.frame %>% rownames_to_column(var="gene_id") %>% 
    pivot_longer(-gene_id, names_to="id", values_to="value") %>% 
    left_join(md) %>% 
    mutate(set= factor(set))
  
  dat_df_avg = dat_df_full %>% group_by(gene_id, position, tags) %>% 
    summarize(value=mean(value)) %>%
    ungroup() 
  
  dat_df_var = dat_df_avg %>%
    group_by(tags, gene_id) %>%
    summarize(value_var_pos = LogVMR(value))  %>%
    group_by(gene_id) %>%
    summarize(value_var_pos_mean = mean(value_var_pos))
  
  
  dat_df_var2 = dat_df_avg %>%
    group_by(position, gene_id) %>%
    summarize(value_var_tags = LogVMR(value)) %>%
    ungroup() %>%
    group_by(gene_id) %>%
    summarize(value_var_tags_mean = mean(value_var_tags))
  
  dat_df_stat = dat_df_var %>% left_join(dat_df_var2) %>%
    filter(is.finite(value_var_pos_mean))
  mod = lm(value_var_pos_mean ~ value_var_tags_mean, dat_df_stat)
  dat_df_stat$resid = residuals(mod)
  saveRDS(dat_df_stat, gene_var_fname)
} else {
  dat_df_stat = readRDS(gene_var_fname)
}
```

```{r}
thresh_pos = .3
thresh_tags = 1
thresh_resid = -1
gg = ggplot(dat_df_stat, aes(value_var_tags_mean, resid)) + 
  geom_point() + 
  geom_hline(yintercept=thresh_resid)
gg

dat_df_stat_thresh = dat_df_stat %>%
  filter(resid > thresh_resid)
```

```{r}
se1 = se1[dat_df_stat_thresh$gene_id,]
```

## Transformations
```{r}
md = se1@meta.data %>% 
  rownames_to_column() %>%
  select(-nsongs_per_day_pre_scale) # from earlier calculation, replacing here

md_tags = md %>%
  distinct(tags, num_songs_on_euth_date, kl_mean, nsongs_per_day_pre, procedure2, duration.of.experiment) %>%
  mutate(num_songs_on_euth_date_log = log1p(num_songs_on_euth_date),
         num_songs_on_euth_date_log_mean = mean(num_songs_on_euth_date_log),
         num_songs_on_euth_date_log_sd = sd(num_songs_on_euth_date_log),
         num_songs_on_euth_date_log_scale = (num_songs_on_euth_date_log - num_songs_on_euth_date_log_mean) / num_songs_on_euth_date_log_sd,
      
         nsongs_per_day_pre = if_else(is.na(nsongs_per_day_pre), mean(nsongs_per_day_pre, na.rm=T), nsongs_per_day_pre),
         nsongs_per_day_pre_log = log1p(nsongs_per_day_pre),
         nsongs_per_day_pre_log_sd = sd(nsongs_per_day_pre_log),
         nsongs_per_day_pre_log_scale = (nsongs_per_day_pre_log - mean(nsongs_per_day_pre_log)) / nsongs_per_day_pre_log_sd ,

         kl_mean_log = log(kl_mean + .1),
         kl_mean_log_scale_cut2_proc2 = cut(kl_mean_log, breaks=c(-3, -1.3, -.65, 2), labels=F),
         kl_mean_log_scale_cut2_proc2 = paste(procedure2, kl_mean_log_scale_cut2_proc2, sep="-"),
         kl_mean_log_scale_cut2_proc2 = factor(kl_mean_log_scale_cut2_proc2, levels=c("intact-1", "intact-2", "deaf-2", "deaf-3"))
  )
```

```{r}
gg= ggplot(md_tags, aes(procedure2, kl_mean_log, color=factor(duration.of.experiment))) + 
  geom_point(position=position_jitter(width = .2)) + 
  geom_hline(yintercept=-.65) + 
  geom_hline(yintercept=-1.3)
gg
```

```{r}
table(md_tags %>% select(procedure2, kl_mean_log_scale_cut2_proc2))

md = md %>% 
  select(-num_songs_on_euth_date, -kl_mean, -nsongs_per_day_pre) %>%
  left_join(md_tags) %>%
  mutate(frac_mito_scale = scale(frac_mito),
         tags_position = paste(tags, position, sep="-")) %>%
  column_to_rownames()

se1 = AddMetaData(se1, md)

#333se1$index2_pool = paste(se1$index2, se1$pool, sep="-")
```

## SVA
```{r}
sv_dir = file.path(prefix_cur, "sva")
dir.create(sv_dir)
sv_fname = file.path(sv_dir, "sva.rds")
vp_fname = file.path(sv_dir, "vp.rds")

dat = as.matrix(GetAssayData(se1, slot = "counts"))
  
redo = F
if (redo) {
  

  
  des_char ="~0 + position +
position:num_songs_on_euth_date_log_scale +
position:kl_mean_log_scale_cut2_proc2 +
position:kl_mean_log_scale_cut2_proc2:num_songs_on_euth_date_log_scale + 
position:nsongs_per_day_pre_log_scale + 
cdr_scaled + frac_mito_scale"
  
  des_char0 = "~0 + position +
cdr_scaled + frac_mito_scale"
  
  mod = model.matrix(as.formula(des_char), se1@meta.data)
  mod0 = model.matrix(as.formula(des_char0), se1@meta.data)
  n.sv = num.sv(dat,mod,method="leek")
  n.sv
  
  svobj = svaseq(dat, mod, mod0)
  sv = svobj$sv
  rownames(sv) = colnames(se1)
  colnames(sv) = paste0("sv", 1:ncol(sv))
  saveRDS(sv, sv_fname)
  
} else {
  sv = readRDS(sv_fname)
}
```

```{r}
library(variancePartition)
redo = F
if (redo) {
  ## Variance partition

  dat_cur = dat
  to_use = rowSums(cpm(dat_cur)>1) >=0.5 * ncol(dat_cur)
  dat_cur1 = dat_cur[to_use,]
  
  info = as.data.frame(cbind(se1@meta.data, sv))
  info = info %>%
    mutate(lib_column = factor(lib_column),
           set = factor(set))
  gExpr = DGEList(counts=dat_cur1)
  gExpr = calcNormFactors(gExpr, method = "TMMwsp")
  
  design = model.matrix( ~ 1, info)
  vobjGenes = voom(gExpr, design )
  
  param = MulticoreParam(workers = 10)
  #param = SnowParam(workers = 10, type = "SOCK")
  
  form = ~ cdr_scaled + frac_mito_scale + sv1 + sv2 + sv3 + sv4 + sv5 + sv6 + sv7 + sv8 + sv9 + sv10 + (1|lib_column) + (1|set) +
    (1|position)
  varPart = fitExtractVarPartModel(vobjGenes, form, info, BPPARAM = param )
  
  
  saveRDS(varPart, vp_fname)
} else {
  varPart = readRDS(vp_fname)
}


vp = sortCols( varPart )
  plotPercentBars( vp[c("CRHBP", "SST", "EGR1", "SLC17A6"),] )
  gg = plotVarPart( vp )
  gg
  save_plot(file.path(sv_dir, "variancePartition_variance.pdf"), gg)
  
vp %>% as.data.frame() %>% rownames_to_column() %>%
  pivot_longer(-rowname) %>%
  group_by(name) %>%
  summarize(value_median = median(value),
            value_mean = mean(value),
            value_n = sum(value>.1)) %>%
  arrange(desc(value_n))

se1 = AddMetaData(se1, as.data.frame(sv))
```


## Model

```{r}
redo = F

max_sv = 2

outdir1 = file.path(prefix_cur, sprintf("sv%s", max_sv))
dir.create(outdir1)

model_fname = file.path(outdir1, "fit.rds")
formula_fname = file.path(outdir1, "formula.txt")         
normalize.method="none"
blocking="tags_position"

se1_cur = se1

des_char = "~0 + position +
position:num_songs_on_euth_date_log_scale +
position:kl_mean_log_scale_cut2_proc2 +
position:kl_mean_log_scale_cut2_proc2:num_songs_on_euth_date_log_scale + 
position:nsongs_per_day_pre_log_scale + 
cdr_scaled + frac_mito_scale"

if (max_sv > 0) {
  svs = paste(paste0("sv", 1:max_sv), collapse=" + ")
  des_char =paste(des_char, svs, sep=" + ")
}

if (redo) {
  fit = run_fit(se1_cur, des_char, blocking)
  saveRDS(fit, model_fname)
  write_lines(des_char, file = formula_fname)
} else {
  fit = readRDS(model_fname)
}

```
### Standard errors
```{r}
fit1 = eBayes(fit)
coefs = fit1$coefficients[,-1]
std_errors = sqrt(fit1$s2.post) * fit1$stdev.unscaled
std_errors = std_errors[,-1]
cd = data.frame(cn = colnames(coefs)) %>%
  filter(grepl("position", cn)) %>% 
  mutate(cn_s = map(cn, ~unlist(str_split(.x, "\\."))),
         cn_s_len = map_int(cn_s, length)) %>%
  filter(cn_s_len > 1) %>%
  mutate(
    position = map_chr(cn_s, 1),
         position = sub("position", "", position),
    coef_orig = sub("position[a-z]+\\.", "", cn))

coefs = coefs[,cd$cn]
std_errors = std_errors[,cd$cn]
se_coefs = SummarizedExperiment(assays=list(coef=coefs, std_errors = std_errors), colData = cd)
```

## Contrasts

```{r}
outdir2 = file.path(outdir1, "by_position")
dir.create(outdir2)
results_fname = file.path(outdir2, "model_results.rds")
redo = F

positions = unique(se1_cur$position)
if (redo) {
  
  design = fit$design
  coefs_names = list(
    kl_mean_2_hearing = c(
      "position%s.kl_mean_log_scale_cut2_proc2intact.2"
    ),
    kl_mean_2_deaf = c(
      "position%s.kl_mean_log_scale_cut2_proc2deaf.2"
    ),
    kl_mean_3 = c(
      "position%s.kl_mean_log_scale_cut2_proc2deaf.3"
    ),
    num_songs_euth_date = c(
      "position%s.num_songs_on_euth_date_log_scale"
      
    ),
    num_songs_pre = c(
      "position%s.nsongs_per_day_pre_log_scale"
      
    ),
    inter = c(
      "position%s.num_songs_on_euth_date_log_scale.kl_mean_log_scale_cut2_proc2deaf.3"
    )
  )
  coefs_names_short = map(coefs_names, ~sub("position%s\\.", "", .x))
  coefs_names_short = coefs_names_short[!grepl("aov", names(coefs_names_short))]
  coefs_names_short_df = data.frame(coef_orig = unlist(coefs_names_short), coef = names(coefs_names_short))
  res = map(positions, function(p) {
    print(p)
    map(coefs_names, function(co) {
      print(co)
      co_cur = sprintf(co, p)
      fit2 = contrasts.fit(fit, coefficients = co_cur)   
      fit2 = eBayes(fit2)
      tt = topTable(fit2, p.value=1, number=nrow(se1), confint=T)
      tt = as.data.frame(tt) %>% rownames_to_column(var="gene_id") %>%
        select(gene_id, AveExpr, P.Value, adj.P.Val, one_of("F", "t", "logFC"))
      tt
    }) %>% set_names(names(coefs_names)) %>% bind_rows(.id="coef")
  }) %>% set_names(positions) %>% bind_rows(.id="position")
  
  std_errors = assay(se_coefs, "std_errors")
  std_errors_df = melt(std_errors)
  colnames(std_errors_df) = c("gene_id", "cn", "std_error")
  se_coefs_md = colData(se_coefs) %>% as.data.frame()
  std_errors_df = std_errors_df %>% left_join(se_coefs_md) %>%
    left_join(coefs_names_short_df)
  res = res %>% left_join(std_errors_df %>% select(gene_id, position, coef, std_error))
  res = res %>% mutate(position = factor(position, levels=position_levels))
  saveRDS(res, results_fname)
} else{
  res = readRDS(results_fname)
}
```



```{r}
tab = res %>% filter(adj.P.Val<.1) %>% group_by(coef, position) %>%
  summarize(n=n()) %>%
  pivot_wider(names_from="coef", values_from="n") %>% arrange(position)

ft = flextable(tab)
ft
save_as_html(ft, path = file.path(outdir2, "deg_table.html"))

tab = res %>% filter(adj.P.Val<.1) %>% 
  mutate(sign = sign(logFC)) %>%
  group_by(coef, position, sign) %>%
  summarize(n=n()) %>%
  pivot_wider(names_from="coef", values_from="n") %>% arrange(position)

ft = flextable(tab)
ft

save_as_html(ft, path = file.path(outdir2, "deg_table_signed.html"))

```
## Write out flat file
```{r}
res_out = res %>% filter(coef %in% c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre"))
flat_fname = file.path(outdir2, "voom_results.csv")
write_csv(res_out, flat_fname)
```

```{r}
md_tags_stats = md_tags %>%
  distinct(num_songs_on_euth_date_log_sd, nsongs_per_day_pre_log_sd) %>%
  pivot_longer(everything(), names_to="coef2", values_to="std.dev") %>%
  mutate(coef2 = sub("_sd", "", coef2))

res = res %>% mutate(coef2 = sub("position%s.", "", coef)) %>%
  mutate(coef2 = sub("_scale", "", coef2)) %>%
  left_join(md_tags_stats)

res = res %>% mutate(logFC_rescale = logFC * std.dev)

res = res %>% left_join(loc_trans)
res = res %>% mutate(gene_id = if_else(is.na(synonym), gene_id, synonym))
```

## MA
```{r}
co ="kl_mean_3"
p = "ra"
res_cur = res %>%
  filter(coef==co, position==p)

gg = ggplot(res_cur, aes(AveExpr, logFC)) + 
  geom_point()+ 
  geom_hline(yintercept=0, color=2)
gg
```

## Volcano

```{r}
volcano_dir = file.path(outdir2, "volcano")
dir.create(volcano_dir)
```


```{r}
params = list(list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
              list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
             )

walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed))
  
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  
  ## Set y max
  res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", color="color_sig")) +
    scale_color_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25) +
    scale_y_continuous(limits=c(0, ymax)) + 
    facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=.1),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("combined_grid_%s_no_text.pdf", co_label)), gg, ncol=2, nrow=4, base_height=1.2, base_aspect_ratio = .8)
})
```

```{r}
params = list(
  list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
  list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
)

walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
    mutate(sig = padjust_signed>1)
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  ## Set y max
  res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  
  sig_label = res_cur %>% 
    filter(sig) %>%
    group_by(position) %>% 
    top_n(-10, wt = adj.P.Val) %>%
      mutate(gene_id_sig = gene_id) %>%
    select(position, gene_id_sig, gene_id)
  res_cur = res_cur %>% left_join(sig_label) %>%
    mutate(point_color = if_else(is.na(gene_id_sig), color_sig, "black"))

  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", fill="color_sig", color="point_color")) +
    scale_color_identity() + 
    scale_fill_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25, shape=21) +
    geom_text_repel(aes(label=gene_id_sig), size=5/2.83, segment.size = .25) + 
    scale_y_continuous(limits=c(0, ymax)) + 
    facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=.1),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1, linetype = "blank"),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("combined_grid_%s_text.pdf", co_label)), gg, ncol=2, nrow=4, base_height=1.2, base_aspect_ratio = .8)
})
```

```{r}
params = list(
  list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
  list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
)

positions = as.character(positions)
walk(positions, function(p) {
  walk(params, function(param) {
    co = param[["co"]]
    xvar = param[["xvar"]]
    ymax = param[["ymax"]]
    xlabel = param[["xlab"]]
    co_label = sub("position%s.", "", co)
    res_cur = res %>% filter(coef==co)
    
    
    res_cur = res_cur %>% 
      filter(position==p) %>%
      mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
      mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
      mutate(sig = padjust_signed>1)
    
    
    info2 = se1@meta.data
    pos_pos2_all = info2 %>%
      distinct(position, position2, position_pretty) %>%
      mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
      mutate(col = position_colors[match(position, names(position_colors))])
    positions = intersect(position_levels, unique(res_cur$position))
    
    
    ## Set y max
    res_cur = res_cur %>% 
      left_join(pos_pos2_all %>% select(position, position2))
    y_max = res_cur %>% 
      group_by(position2) %>%
      summarize(y_max = max(padjust_signed))
    res_cur = res_cur %>%
      left_join(y_max)
    
    res_cur = res_cur %>% left_join(pos_pos2_all) %>%
      mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
      mutate(position = factor(position, levels=position_levels),
             position_pretty = factor(position_pretty, levels=position_pretty_levels))
    
    
    sig_label = res_cur %>% 
      filter(sig) %>%
      group_by(position) %>% 
      top_n(-10, wt = adj.P.Val) %>%
      mutate(gene_id_sig = gene_id) %>%
      select(position, gene_id_sig, gene_id)
    res_cur = res_cur %>% left_join(sig_label) %>%
      mutate(point_color = if_else(is.na(gene_id_sig), color_sig, "black"))
    
    
    gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", fill="color_sig", color="point_color")) +
      scale_color_identity() + 
      scale_fill_identity() + 
      geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
      geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
      geom_point(size = .25, shape=21) +
      geom_text_repel(aes(label=gene_id_sig), size=5/2.83, segment.size = .25) + 
      scale_y_continuous(limits=c(0, ymax)) + 
      labs(x=xlabel,
           y="-log10(adjusted p-value)") + 
      theme_bw() + 
      theme(legend.position="none",
            axis.title = element_text(size=6),
            axis.text = element_text(size=5),
            axis.line = element_line(size=0),
            axis.ticks  = element_line(size=.1),
            strip.background = element_rect(size=1, linetype = "blank"),
            strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
            panel.grid = element_blank()
      ) 
    gg
    save_plot(file.path(volcano_dir, sprintf("%s_%s_text.pdf", p, co_label)), gg, ncol=1, nrow=1, base_height=1.2, base_aspect_ratio = .8)
  })
})
```
###### Neuropeptide
Each region separately
```{r}
params = list(list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
              list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
             )

walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
        filter(gene_id %in% nps$SYMBOL)
  
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  
  ## Set y max
  res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", color="color_sig")) +
    scale_color_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25) +
    scale_y_continuous(limits=c(0, ymax)) + 
    facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=.1),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("combined_grid_%s_nps_no_text.pdf", co_label)), gg, ncol=2, nrow=4, base_height=1.2, base_aspect_ratio = .8)
})
```

```{r}
params = list(list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
              list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance"),
              list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
             )

walk(params, function(param) {
  co = param[["co"]]
  xvar = param[["xvar"]]
  ymax = param[["ymax"]]
  xlabel = param[["xlab"]]
  co_label = sub("position%s.", "", co)
  res_cur = res %>% filter(coef==co)
  
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))  %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
    filter(gene_id %in% nps$SYMBOL)
  
  res_cur = res_cur %>% mutate(padjust_signed = -1 * log10(adj.P.Val))
  
  if (is.null(ymax))
    ymax = max(res_cur$padjust_signed)
  
  res_cur = res_cur %>%
    mutate(padjust_signed = if_else(padjust_signed>ymax, ymax, padjust_signed)) %>%
    mutate(sig = padjust_signed>1)
  
  info2 = se1@meta.data
  pos_pos2_all = info2 %>% distinct(position, position2, position_pretty) %>%
    mutate(songsystem = as.integer(position %in% c("ra", "hvc", "lman", "x"))) %>%
    mutate(col = position_colors[match(position, names(position_colors))])
  positions = intersect(position_levels, unique(res_cur$position))
  
  
  ## Set y max
  res_cur = res_cur %>% left_join(pos_pos2_all %>% select(position, position2))
  y_max = res_cur %>% 
    group_by(position2) %>%
    summarize(y_max = max(padjust_signed))
  res_cur = res_cur %>%
    left_join(y_max)
  
  res_cur = res_cur %>% left_join(pos_pos2_all) %>%
    mutate(color_sig = if_else(padjust_signed>1, col, "grey50")) %>%
    mutate(position = factor(position, levels=position_levels),
           position_pretty = factor(position_pretty, levels=position_pretty_levels))
  
  sig_label = res_cur %>% 
    filter(sig) %>%
    group_by(position) %>% 
    #top_n(-10, wt = adj.P.Val) %>%
      mutate(gene_id_sig = gene_id) %>%
    select(position, gene_id_sig, gene_id)
  res_cur = res_cur %>% left_join(sig_label) #%>%
  
  gg = ggplot(res_cur, aes_string(xvar, "padjust_signed", color="color_sig")) +
    scale_color_identity() + 
    geom_hline(yintercept=1, linetype=1, size=.25, color="grey") + 
    geom_vline(xintercept=0, linetype=1, size=.25, color="grey") + 
    geom_point(size = .25) +
     geom_text_repel(aes(label=gene_id_sig), size=5/2.83, segment.size = .25) + 
    scale_y_continuous(limits=c(0, ymax)) + 
    facet_wrap(~position_pretty, ncol=2) +
    labs(x=xlabel,
         y="-log10(adjusted p-value)") + 
    theme_bw() + 
    theme(legend.position="none",
          axis.title = element_text(size=6),
          axis.text = element_text(size=5),
          axis.line = element_line(size=.1),
          axis.ticks  = element_line(size=.1),
          strip.background = element_rect(size=1),
          strip.text = element_text(size=6, face="bold", margin = margin(1,0,1,0)),
          panel.grid = element_blank()
    ) 
  gg
  save_plot(file.path(volcano_dir, sprintf("combined_grid_%s_nps_text.pdf", co_label)), gg, ncol=2, nrow=4, base_height=.8, base_aspect_ratio = 1)
})
```


### DEG score
```{r}
deg_dir = file.path(outdir2, "deg_scores")
dir.create(deg_dir)
```

```{r}
params = list(
  list(co = "kl_mean_3", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")#,
 # list(co = "kl_mean_2_deaf", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")#,
#       list(co = "kl_mean_2_hearing", xvar = "logFC", ymax = 5, xlab = "log fold-change mean K-L distance")
  )
  
  res_sum = res %>%
  mutate(padjust_signed = -1 * log10(adj.P.Val)) %>%
  filter(padjust_signed>1) %>%
  mutate(sign = sign(logFC)) %>%
  ungroup() %>% 
  group_by(coef, position, sign) %>%
  summarize(score = sum(padjust_signed)) %>%
  ungroup() %>% 
  mutate(score_sign = score * sign) %>%
  mutate(position_sign = ifelse(sign>0, as.character(position), NA))

coefs_to_plot = map_chr(params, "co")
res_sum_to_plot = res_sum %>% filter(coef %in% coefs_to_plot )
min_y = min(res_sum_to_plot$score_sign)
max_y = max(res_sum_to_plot$score_sign)

walk(params, function(param) {
  co = param[["co"]]
  
co_label = sub("position%s.", "", co)
res_cur = res_sum %>% 
  filter(coef==co) %>%
  mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
  complete(position, fill=list(score = 0, score_sign = 0))

gg = ggplot(res_cur, aes(position, score_sign)) + 
  geom_bar(stat="identity", aes(color=position, fill=position_sign), size=.25) + 
  scale_fill_manual(name = "",values=position_colors, guide=F, na.value=NA) + 
  scale_color_manual(values=position_colors, guide=F) + 
  ylim(min_y, max_y) + 
  labs(y="DEG score", title=co) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid.major.x=element_blank(),  
        panel.grid.minor.y = element_blank())
print(gg)
save_plot(file.path(deg_dir, sprintf("%s.pdf" ,co_label)), gg, base_height = 1, base_width=1.2)
})
```

```{r}
params = list(
  list(co = "num_songs_euth_date", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co = "num_songs_pre", xvar = "logFC", ymax = 5, xlab = "log fold-change # songs on euthanasia date"),
  list(co=  "inter", xvar = "logFC", ymax=5, xlab="log fold-change singing x song destabilization")
)

res_sum = res %>%
  mutate(padjust_signed = -1 * log10(adj.P.Val)) %>%
  filter(padjust_signed>1) %>%
  mutate(sign = sign(logFC)) %>%
  ungroup() %>% 
  group_by(coef, position, sign) %>%
  summarize(score = sum(padjust_signed)) %>%
  ungroup() %>% 
  mutate(score_sign = score * sign) %>%
  mutate(position_sign = ifelse(sign>0, as.character(position), NA))

coefs_to_plot = map_chr(params, "co")
res_sum_to_plot = res_sum %>% filter(coef %in% coefs_to_plot )
min_y = min(res_sum_to_plot$score_sign)
max_y = max(res_sum_to_plot$score_sign)

walk(params, function(param) {
  co = param[["co"]]
  
co_label = sub("position%s.", "", co)
res_cur = res_sum %>% 
  filter(coef==co) %>%
  mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
  complete(position, fill=list(score = 0, score_sign = 0))

gg = ggplot(res_cur, aes(position, score_sign)) + 
  geom_bar(stat="identity", aes(color=position, fill=position_sign), size=.25) + 
  scale_fill_manual(name = "",values=position_colors, guide=F, na.value=NA) + 
  scale_color_manual(values=position_colors, guide=F) + 
  ylim(min_y, max_y) + 
  labs(y="DEG score", title=co) + 
  theme_bw() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid.major.x=element_blank(), 
        panel.grid.minor.y = element_blank())
print(gg)
save_plot(file.path(deg_dir, sprintf("%s.pdf" ,co_label)), gg, base_height = 1, base_width=1.2)
}) 
```

## Heatmap

```{r}
heat_dir = file.path(outdir2, "heatmaps")
dir.create(heat_dir)
```

```{r}

positions_to_use = c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri")
coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3", "num_songs_euth_date", "num_songs_pre", "inter")
cols = list("RdBu", "PuOr", "BrBG", "RdBu")
names(cols) = coefs_to_use

res_mean = res %>% group_by(coef, gene_id) %>% 
  filter(gene_id %in% nps$SYMBOL)
walk(coefs_to_use, function(co) {
  print(co)
  res_sig = res %>%
    filter(gene_id %in% nps$SYMBOL) %>%
    filter(coef==co) %>% 
    group_by(position) %>%
    filter(adj.P.Val<.1)# %>%
    #filter(logFC_abs > .58)
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  #value_var = ifelse(co %in% c("num_songs_euth_date", "position%s.kl_mean_log_scale" ), "logFC_rescale", "logFC")
  value_var = "logFC"
  res_cur = res %>% 
    filter(gene_id %in% res_sig$gene_id) %>%
    filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale)
              )
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  #colnames(mat) = toupper(colnames(mat))
  #mat = mat[,positions_to_use]
  
  #mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = 1
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  length_out = 9
  cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[1]])(length_out)))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("heatmap_np_sig_%s.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(heat_dir, sprintf("heatmap_np_sig_%s_horiz.pdf", co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  hm_t
})
```

```{r}

coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre")
cols = list("RdBu", "PuOr", "BrBG", "RdBu")
names(cols) = coefs_to_use

# res_mean = res %>% group_by(coef, position, gene_id) %>% 
#   summarize()
walk(coefs_to_use, function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    #group_by(position) %>%
    filter(adj.P.Val<.1) #%>%
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  

  value_var = "logFC"
  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale))
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  mat = mat[,c("ra", "hvc", "lman", "x")]
  colnames(mat) = toupper(colnames(mat))
  
  mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .25
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  #length_out = 9
 # cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
  cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("blue"), "white", muted("red")))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_song-only.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_song-only_horiz.pdf", co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```


```{r}
coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre")
cols = list("RdBu", "PuOr", "BrBG", "RdBu")
names(cols) = coefs_to_use

#res_mean = res %>% group_by(coef, position, gene_id) %>% 
#  summarize()


res = res %>%
  mutate(songsystem = case_when(position %in% c("ra", "hvc", "lman", "x") ~ TRUE,
                                TRUE ~ FALSE),
         position2 = case_when(position %in% c("ra", "arco") ~ "ra",
                               position %in% c("hvc", "ncl") ~ "hvc",
                               position %in% c("lman", "nido") ~ "lman",
                               position %in% c("x", "stri") ~ "x"))

walk(coefs_to_use, function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    #group_by(position) %>%
    filter(adj.P.Val<.1) #%>%
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  
  
  #value_var = ifelse(co %in% c("num_songs_euth_date", "position%s.kl_mean_log_scale" ), "logFC_rescale", "logFC")
  value_var = "logFC_rel"
  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>%
    filter(coef==co) %>%
    group_by(position, gene_id, position2, songsystem) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale)) %>%
    group_by(position2, gene_id) %>%
    mutate(logFC_rel = logFC - logFC[!songsystem])
  
  mat = acast(res_cur, gene_id~position, value.var=value_var)
    mat = mat[,c("ra", "hvc", "lman", "x")]
  colnames(mat) = toupper(colnames(mat))

  #mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = .25
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  length_out = 9
 #cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[co]])(length_out)))
    cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("blue"), "white", muted("red")))
  hc_row = hclust(dist(mat, method="euclidean"), method = "ward.D2")
  
  hm = Heatmap(mat, col=cols_cur,
               cluster_rows = hc_row,
               cluster_columns = F,
               width = unit(width, "in"),
               height = unit( height, "in"),
               column_names_gp = gpar(fontsize=5), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=5)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_norm-surround.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(heat_dir, sprintf("heatmap_%s_top_fc_norm-surround_horiz.pdf", co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  #hm
})
```

```{r}

walk(unique(res$coef), function(co) {
  print(co)
  res_sig = res %>%
    filter(coef==co) %>% 
    group_by(position) %>%
        filter(adj.P.Val<.1)  %>%
    top_n(-20, adj.P.Val)

  
  value_var = "t"
  mat = acast(res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co), gene_id~position, value.var=value_var)
  colnames(mat) = toupper(colnames(mat))
  
  #mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .5),]
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  
  hm = Heatmap(mat, 
               clustering_method_rows = "average", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_t_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  hm
})
```

```{r}
walk(unique(res$coef), function(co) {

  res_sig = res %>%
    filter(position %in% c("ra", "hvc", "lman", "x")) %>% 
    filter(coef==co) %>% 
    group_by(position) %>%
    filter(adj.P.Val<.1) %>%
    top_n(10, abs(logFC))

  res_cur = res %>% filter(gene_id %in% res_sig$gene_id) %>% filter(coef==co) %>%
     filter(position %in% c("ra", "hvc", "lman", "x"))
  mat = acast(res_cur, gene_id~position, value.var="logFC")
  colnames(mat) = toupper(colnames(mat))
  
  width = .75
  height = width * nrow(mat) / ncol(mat)
  
  
  hm = Heatmap(mat, clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(heat_dir, sprintf("%s_top_fc_songsystem_heatmap.pdf", co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  hm
})
```

## GSEA
```{r}
gsea_dir = file.path(outdir2, "gsea")
dir.create(gsea_dir)
```

```{r}
gmt_dir = file.path(base_dir, "deaf_gex", "seq_analysis", "common_data")
gmt_files = list.files(gmt_dir, full.names = T)
gmt_files = gmt_files[grep("c5.all.v7", gmt_files)]

gmt_pathway = gmtPathways(gmt_files)
to_exclude = names(gmt_pathway)
to_exclude = to_exclude[grep("ENDOPLAS|RIBOSOM|TRANSLATION|METABO|CATABOL|RESPIR|HYDROGEN|OXIDAT|OXIDO|MITOCH|PROTON|VIRAL|ELECTRON|ATP|RRNA|BIOSYNTHETIC|INORGANIC|INNER_MEMBRANE", to_exclude)] ## Large variation in associated genes across indivduals
gmt_pathway = gmt_pathway[!(names(gmt_pathway) %in% to_exclude)]

gray_gmt_fname = file.path(gmt_dir, "gray_2018_arg.gmt") ## Obtained from Tyssowski, K. M. et al. Different Neuronal Activity Patterns Induce Different Gene Expression Programs. Neuron 98, 530–546.e11 (2018)
gray_gmt = gmtPathways(gray_gmt_fname)
names(gray_gmt) = make.names(names(gray_gmt))
gmt_gray_pathway = c(gmt_pathway, gray_gmt)
```

```{r}
positions = as.character(unique(res$position))
coefs = unique(res$coef)

plan(multisession, workers = length(positions))
options(future.globals.maxSize = 1024^2 * 5000)
```

```{r message=FALSE, warning=FALSE}

gsea_res_fname = file.path(gsea_dir, "gsea_res.qs")
gsea_res_ind_fname = file.path(gsea_dir, "gsea_res_ind.qs")
gsea_res_flat_fname = file.path(gsea_dir, "gsea_res.csv")
redo = F
if (redo) {
  positions = unique(res$position)
  
  gsea_res = future_map(positions, function(p) {
    print(p)
    res_cur = res %>% filter(position == p)
    map(coefs, function(co) {
      print(p)
      res_cur1 = res_cur  %>% filter(coef==co)
      rnks = res_cur1$t
      names(rnks) = res_cur1$gene_id
      rnks = sort(rnks)
      gsea_res = fgseaMultilevel(pathways = gmt_gray_pathway,
                                 stats = rnks,
                                 minSize=20,
                                 maxSize=200)
      return(gsea_res)
      
    }) %>% set_names(coefs) %>% bind_rows(.id="coef")
  })  %>% set_names(positions) %>% bind_rows(.id="position")
  
  gsea_res_ind = future_map(positions, function(p) {
    print(p)
    res_cur = res %>% filter(position == p)
    gsea_res_cur = gsea_res %>% filter(position==p)
    map(coefs, function(co) {
      res_cur1 = res_cur  %>% filter(coef==co)
      gsea_res_cur1 = gsea_res_cur %>% filter(coef==co)
      rnks = res_cur1$t
      names(rnks) = res_cur1$gene_id
      rnks = sort(rnks)
      gsea_res_filt = gsea_res_cur1 %>% filter(padj<.2) %>% arrange(pval)
      collapsedPathways = collapsePathways(gsea_res_filt, pathways = gmt_gray_pathway, stats = rnks, pval.threshold = .05)
      gsea_res_ind = gsea_res_cur1 %>% filter(pathway %in% collapsedPathways$mainPathways) %>%
        arrange(-NES)
      gsea_res_ind
    }) %>% set_names(coefs) %>% bind_rows(.id="coef")
  }) %>% bind_rows()
  
  qsave(gsea_res, gsea_res_fname)
  qsave(gsea_res_ind, gsea_res_ind_fname)
} else {
  gsea_res = qread(gsea_res_fname)
  gsea_res_ind = qread(gsea_res_ind_fname)
}

```

### Write out flat file
```{r}
gsea_res_ind_out = gsea_res_ind %>% filter(coef %in% c("kl_mean_3", "num_songs_euth_date", "inter", "num_songs_pre"))
flat_fname = file.path(gsea_dir, "gsea_results.csv")
write_csv(gsea_res_ind_out, flat_fname)
```

###### Heatmap
```{r}
coefs = c("kl_mean_2_hearing", "kl_mean_2_deaf", "kl_mean_3", "num_songs_euth_date", "num_songs_pre", "inter")
walk(coefs, function(co) {
  print(co)
  co_label = sub("position%s.", "", co)
  gsea_res_ind_filt = gsea_res_ind %>% 
    filter(padj<.1) %>%
    filter(coef==co)
  
  gsea_res_filt = gsea_res %>% 
    mutate(padj_signed = sign(NES) *  -1 * log10(padj)) %>% 
    filter(pathway %in% gsea_res_ind_filt$pathway)
  
  mat = gsea_res_filt %>% filter(coef==co) %>%
    select(position, pathway, padj_signed) %>%
    pivot_wider(names_from = "position", values_from = "padj_signed", values_fill=list(NES=0)) %>%
    column_to_rownames("pathway") %>% 
    as.matrix() 
  
  mat = mat[,intersect(position_levels, colnames(mat))]
  
  max_val = 1.5
  cols_cur = colorRamp2(breaks=c(-1*max_val, 0, max_val), colors = c(muted("purple"), "white", muted("orange")))
  
  width = .5
  height = 2.5
  hm = Heatmap(mat, 
               cluster_columns=F, col = cols_cur,
               clustering_method_rows = "ward.D",
               width = unit(width, "in"),
               height = unit(height, "in"))
  hm
  
  pdf(file.path(gsea_dir, sprintf("heatmap_pval_%s.pdf", co_label)), width=width + 8, height=height + 2)
  print(hm)
  dev.off()
  
  width = 1
  height = width * nrow(mat) / ncol(mat)
  
  hm = Heatmap(mat, 
               cluster_columns=F, col = cols_cur,
               clustering_method_rows = "ward.D",
               width = unit(width, "in"),
               height = unit(height, "in"))
  hm
  
  pdf(file.path(gsea_dir, sprintf("heatmap_pval_%s_big.pdf", co_label)), width=width + 8, height=height + 2)
  print(hm)
  dev.off()
})
```

```{r}
coefs_to_use = list(c("kl_mean_3"))
walk(coefs_to_use, function(coef_to_use) { 
  gsea_res_ind_select = gsea_res_ind %>% 
    filter(coef %in% coef_to_use) %>% 
    mutate(p.adjust_sign = -1 * log10(padj)) %>%
    mutate(NES_sign = sign(NES)) %>%
    filter(padj<.1) %>% 
    group_by(position, NES_sign) %>%
    top_n(5, abs(NES))
  
  gsea_res_ind_filt = gsea_res_ind %>% 
    filter(coef %in% coef_to_use) %>% 
    filter(pathway %in% gsea_res_ind_select$pathway) %>%
    distinct(position, pathway, .keep_all=T) %>%
    mutate(pathway = sub("^GO", "", pathway ),
           pathway = gsub("_", " ", pathway)) %>% 
    mutate(pathway = case_when(grepl("NONSENSE MEDIATED DECAY",pathway) ~ "NONSENSE MEDIATED DECAY",
                               grepl("SERINE THREONINE KINASE", pathway) ~ "SERINE THREONINE KINASE SIGNALING",
                               TRUE ~ pathway))
  
  go_res_mat = gsea_res_ind_filt %>% select(position, pathway, NES) %>%
    pivot_wider(names_from = "position", values_from = "NES", values_fill = list(NES= 0)) %>%
    column_to_rownames("pathway") %>% as.matrix()
  
  hc_row = hclust(dist(go_res_mat), method="complete")
  hc_row_order = hc_row$labels[hc_row$order]
  
  
  gsea_res_ind_filt = gsea_res_ind_filt %>% 
    mutate(leadingEdge_len = map_int(leadingEdge, length))
  
  gsea_res_ind_filt = gsea_res_ind_filt %>%
    mutate(pathway = factor(pathway, levels=hc_row_order)) %>%
    mutate(position = factor(position, levels=c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri"))) %>%
    complete(pathway, position, fill=list(NES=NA, size=NA, leadingEdge_len = NA))
  
  size_limits = seq(round(min(gsea_res_ind_filt$size, na.rm=T), -1), round(max(gsea_res_ind_filt$size, na.rm=T), -1), by=50)
  size_limits = seq(round(min(gsea_res_ind_filt$leadingEdge_len, na.rm=T), -1), round(max(gsea_res_ind_filt$leadingEdge_len, na.rm=T), -1), by=20)
  gg = ggplot(gsea_res_ind_filt, aes(position, pathway)) + 
    geom_point(aes(size=leadingEdge_len, color=NES)) + 
    scale_color_gradient2(low=muted("blue"), high=muted("red")) +
    scale_size_continuous(range=c(.5, 3), breaks=size_limits) + 
    labs(x = "", y ="") + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1, size=6),
          axis.text.y = element_text(hjust=1, size=6),
          axis.ticks = element_line(size=.25),
          panel.border = element_blank()) + 
    coord_fixed()
  print(gg)
  ncols = ncol(go_res_mat)
  nrows = nrow(go_res_mat)
  
  width = ncols / 10
  height = nrows / 10
  gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
  save_plot(file.path(gsea_dir, sprintf("go_terms_by_module_%s.pdf", paste(coef_to_use, collapse="-"))), gg, base_height = (nrows / 2) + 4, base_width = (ncols / 2) + 4)
})
```



###### Plot leading edges

```{r}
pathway_to_use = "GO_HORMONE_ACTIVITY"
p = "ra"

positions_to_use = c("ra", "arco", "hvc", "ncl", "lman", "nido", "x", "stri")
coefs_to_use = unique(res$coef)
coefs_to_use = c("kl_mean_3")
cols = list("RdBu")
names(cols) = coefs_to_use

walk(coefs_to_use, function(co) {
  genes_select = gsea_res_ind %>% filter(position==p, pathway==pathway_to_use, coef==co) %>% select(leadingEdge) %>% unlist()
  print(co)
  res_sig = res %>%
    filter(gene_id %in% genes_select)
    #filter(logFC_abs > .58)
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  #value_var = ifelse(co %in% c("num_songs_euth_date", "position%s.kl_mean_log_scale" ), "logFC_rescale", "logFC")
  value_var = "logFC"
  res_cur = res %>% 
    filter(gene_id %in% res_sig$gene_id) %>%
    filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale)
              )
  mat = acast(res_cur, gene_id~position, value.var=value_var)
  #colnames(mat) = toupper(colnames(mat))
  #mat = mat[,positions_to_use]
  
  #mat_var = apply(mat, 1, var)
  #mat = mat[mat_var>quantile(mat_var, .25),]
  width = 1
  height = width * nrow(mat) / ncol(mat)
  
  max_val = max(c(abs(min(mat)), max(mat)))
  max_val = 2
  length_out = 9
  cols_cur = colorRamp2(breaks=seq(-1*max_val, max_val, length.out=length_out), colors = rev(brewer_pal(palette=cols[[1]])(length_out)))
  hm = Heatmap(mat, col=cols_cur,
               clustering_method_rows = "ward.D", 
               cluster_columns = F,
               width = unit(width, "in"),
               height =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
    co = sub("\\%s", "", co)
  pdf(file.path(gsea_dir, sprintf("gsea_heatmap_%s_%s_%s_leadingEdge.pdf", pathway_to_use, p, co)), width=width+2, height=height+2)
  
  print(hm)
  dev.off()
  #hm
  
  hm_t = Heatmap(t(mat), col=cols_cur,
               clustering_method_columns  = "average", 
               cluster_rows = F,
               height = unit(width, "in"),
               width =unit( height, "in"),
               column_names_gp = gpar(fontsize=6), column_names_side = "top",
               row_names_gp = gpar(fontsize=6)
               #  clustering_method_columns = "ward.D"
  )
  
    pdf(file.path(gsea_dir, sprintf("gsea_heatmap_%s_%s_%s_leadingEdge_horiz.pdf", pathway_to_use, p, co)), width=height+2, height=width+2)
  print(hm_t)
  dev.off()
  hm_t
})
```
```{r fig.height=10, fig.width=5}
genes_select = gsea_res_ind %>% filter(position==p, pathway==pathway_to_use, coef==co) %>% select(leadingEdge) %>% unlist()
genes_select = c(genes_select, "CRHBP")
res_sig = res %>%
    filter(gene_id %in% genes_select)
    #filter(logFC_abs > .58)
    #top_n(50, abs(logFC))
    #top_n(-50, adj.P.Val)
  
  #value_var = ifelse(co %in% c("num_songs_euth_date", "position%s.kl_mean_log_scale" ), "logFC_rescale", "logFC")
  value_var = "logFC"
  res_cur = res %>% 
    filter(gene_id %in% res_sig$gene_id) %>%
    filter(coef==co) %>%
    group_by(position, gene_id) %>%
    summarize(logFC = mean(logFC),
              logFC_rescale = mean(logFC_rescale)
              )
  
gg = ggplot(res_cur, aes(position, logFC, fill=position)) + 
  geom_bar(stat="identity") + 
  scale_fill_manual(values=position_colors) + 
  geom_hline(yintercept=0) +
  theme_bw() + 
  facet_grid(gene_id~.) + 
  geom_text(aes(x=1, y=1, label=gene_id), size = 2, fontface = "italic", hjust=0) + 
  labs(x="", y = "log fold-change mean K-L distance") + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=6),
        axis.ticks.x = element_blank(),
        strip.text = element_blank(),
        legend.position="none",
        panel.grid.major.x=element_blank(),
        axis.title = element_text(size=7))
gg
height = length(genes_select) * .5
save_plot(file.path(gsea_dir, sprintf("gsea_bar_%s_%s_%s_leadingEdge_horiz.pdf", pathway_to_use, p, co)), gg, base_height=height, base_asp = .3)
```

```{r}
pathways_to_use = c(
  "GO_GLIAL_CELL_DIFFERENTIATION",
  "GO_NEURON_SPINE",
  "GO_HORMONE_ACTIVITY",
  "GO_POSITIVE_REGULATION_OF_NEURON_DEATH",
  "GO_PRIMARY_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY",
  "GO_DNA_BINDING_TRANSCRIPTION_ACTIVATOR_ACTIVITY",
  "GO_BRANCHING_MORPHOGENESIS_OF_AN_EPITHELIAL_TUBE"
)
coef_to_use = c("kl_mean_3")

nrows = 5
positions = as.character(positions)
walk(positions, function(p) {
  walk(pathways_to_use, function(pathway_to_use) {
    print(pathway_to_use)
    gsea_res_ind1 = gsea_res %>% 
      filter(pathway==pathway_to_use) %>%
      filter(coef==coef_to_use) %>%
      filter(position==p) 
    genes = unique(gsea_res_ind1$leadingEdge[[1]])[1:20]
    
    line_color = position_colors[p]
    
    res_select = res %>% 
      filter(coef == coef_to_use)  %>%
      filter(position == p) %>%
      filter(gene_id %in% genes) %>%
      mutate(coef = factor(coef, levels=c("baseline", coef_to_use))) %>%
      complete(coef, gene_id, fill=list(logFC=0, std_error=0))
    
    df_group_sum = res_select  %>%
      group_by(coef) %>%
      summarize_at(vars(logFC,std_error), list(mean = mean)) %>%
      ungroup()
    
    gg = ggplot() + 
      geom_hline(yintercept=0, linetype=1, size=.25) + 
      geom_line(data=res_select, color="grey", alpha=.5, aes(coef, logFC,  group=gene_id), size=.25) +
      geom_line(data=df_group_sum, color=line_color, aes(coef, logFC_mean, group=1)) +
      labs(x = "", y = "logFC") + 
      theme_bw() + 
      theme(
        legend.position="none",
        strip.text = element_text(size=6, face="italic", hjust=0, vjust=1),
        strip.background = element_blank(),
        axis.title.y = element_text(size=6),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.text.x = element_blank(),
        axis.line = element_blank(),
        axis.ticks.y = element_line(size=.25),
        axis.ticks.x = element_blank(),
        panel.grid=element_blank()
      )
    print(gg)
    width = .5
    height = .6
    gg = set_panel_size(gg, width = unit(width, "in"), height = unit(height, "in"))
    save_plot(file.path(gsea_dir, sprintf("points_by_kl_mean_log_scale_cut2_proc2-low-hi_logFC_%s_%s_grouped.pdf", p, pathway_to_use)), gg, base_height=2, base_width=2, ncol=1, nrow=1)
    
  })
})

```
